reps/2023-12-14-virtual-cluster.md

```diff
@@ -0,0 +1,807 @@
+```
+
+```python
+# Create a placement group with two bundles that are strict spreaded
+pg = placement_group([{"GPU": 4}, {"GPU": 1}], strategy="STRICT_SPREAD")
+# Run the actor using the bundle 0 resources
+Actor.options(
+    scheduling_strategy=PlacementGroupSchedulingStrategy(
+        placement_group=pg, placement_group_bundle_index=0)).remote()
+
+# Create a virtual cluster with two virtual nodes that are strict spreaded
+vc = VirtualCluster(
+    fixed_size_nodes=[
+        FixedSizeNodes(
+            nodes=[
+                FixedSizeNode(resources={"GPU": 4}, labels={"bundle_index": "0"}),
+                FixedSizeNode(resources={"GPU": 1}, labels={"bundle_index": "1"})
+            ],
+            scheduling_policy=STRICT_SPREAD
+        )
+    ])
+with vc:
+  # Run the actor using the first virtual node resources
+  Actor.options(node_labels={"bundle_index": In("0")}).remote()
+```
+
+Here we have two options for the API:
+
+1. Deprecate the placement group API and use virtual cluster API directly.
+2. Keep the placement group API and only change the internal implementation to use virtual cluster.
+
+#### Examples
+
+In this section, we show several Ray workloads and how they can be built on top of virtual clusters.
+For each example, we show both the code using current Ray APIs and the code using the new virtual cluster APIs.
+
+##### Gang scheduling a group of actors
+
+```python
+ray.init()
+pg = placement_group([{"GPU": 1}, {"GPU": 1}], strategy="STRICT_SPREAD")
+actors = []
+for i in range(2):
+    actors.append(Actor.options(
+        num_gpus=1,
+        scheduling_strategy=PlacementGroupSchedulingStrategy(
+            placement_group=pg, placement_group_bundle_index=i)).remote())
+```
+
+```python
+ray.init(virtual_cluster=VirtualCluster(
+    fixed_size_nodes=[
+        FixedSizeNodes(
+            nodes=[
+                FixedSizeNode(resources={"GPU": 1}, labels={"bundle_index": "0"}),
+                FixedSizeNode(resources={"GPU": 1}, labels={"bundle_index": "1"})
+            ],
+            scheduling_policy=STRICT_SPREAD
+        )
+    ]))
+
+actors = []
+for i in range(2):
+    actors.append(Actor.options(
+        num_gpus=1,
+        node_labels={"bundle_index": In(str(i))}).remote())
+```
+
+##### Tune + Dataset
+
+```python
+# Tune can run 2 trials (each trial runs two 1 GPU trainers) in parallel and Dataset can use 1-10 CPUs
+ray.init(virtual_cluster=VirtualCluster(
+    fixed_size_nodes=[
+        FixedSizeNodes(
+            nodes=[
+                FixedSizeNode(resources={"GPU": 1}),
+                FixedSizeNode(resources={"GPU": 1})
+            ],
+            scheduling_policy=PACK
+        ),
+        FixedSizeNodes(
+            nodes=[
+                FixedSizeNode(resources={"GPU": 1}),
+                FixedSizeNode(resources={"GPU": 1})
+            ],
+            scheduling_policy=PACK
+        )
+    ],
+    flex_resource_min={"CPU": 1},
+    flex_resource_max={"CPU": 10}))
```

Comment:
Trial 1: {fixed_nodes={2 GPU nodes}, flex={2..20 CPUs}}
Trial 2: {fixed_nodes={2 GPU nodes}, flex={2..20 CPUs}}

For the SplitCoordinator, I'm assuming you mean there is some separate preprocessing job that's run here before kicking off the training trials? In that case, you can put that in a another VC with {flex={1-10 CPU, 1-4 GPU}}.

Reply:
The `SplitCoordinator` I'm mentioning is the thing executing the ray data + ray train streaming execution that gets run in the background during the training execution.

https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/iterator/stream_split_iterator.py#L47

Reply:
I see, then those should be part of the trial cluster, so this would be a reasonable config:

Trial 1: {fixed_nodes={2 GPU nodes}, flex={2..20 CPUs}}
Trial 2: {fixed_nodes={2 GPU nodes}, flex={2..20 CPUs}}

For more predictable performance you'd probably also want to assign a fixed number of CPUs instead of allowing 2..20 for flex, so probably `{fixed_nodes={2 GPU nodes}, flex={10 CPUs}}` would be the way to go.

Reply:
```
The Ray Data components are currently detached from the Ray Train/Tune placement group. How does [placement_group_capture_child_tasks=None](https://docs.ray.io/en/latest/ray-core/scheduling/placement-group.html#advanced-child-tasks-and-actors) translate to VirtualClusters?
```

To answer this question directly, by design we don't need to support detached resources for virtual clusters. The data tasks can use flex resources instead.